### PR TITLE
New Kitex blog

### PR DESCRIPTION
**Pull request by yy2so**
New Kitex blog

### CODE REVIEW

**Comment by GuangmingLuo** [comment]
@joway @felix021  plz confirm it

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 1]
```diff
@@ -0,0 +1,538 @@
+---
+date: 2024-01-18
+title: "Enhancing Performance in Microservice Architecture with Kitex"
+projects: ["CloudWeGo"]
+linkTitle: "Enhancing Performance in Microservice Architecture with Kitex"
+keywords: ["CloudWeGo", "middleware", "Go", "Golang", "Kitex", "microservice framework", "ByteDance Open Source", "ByteDance", "open source", "cloud native", "open source", "kubernetes", "gRPC", "microservices", "rpc", "thrift"]
+description: "Volo is a Rust Framework for cloud development. This guide provides insights into leveraging Volo with its speed and efficiency in your projects."
```
Comment (GuangmingLuo) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:7]: This is not correct

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 2]
```diff
@@ -0,0 +1,538 @@
...
+---
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/1.jpeg)
+
+# Kitex: Enhancing Performance in Microservice Architecture
+## Introduction
+
+The team at ByteDance initiated the creation of the Remote Procedure Call (RPC) framework, Kitex, alongside several related fundamental libraries in 2019. This endeavor originated from confronting functionality and performance challenges within our extensive microservice architecture. We also wanted to combine the knowledge and insights gathered from previous frameworks. This development project was officially released for open-source contribution on GitHub in 2021. 
+
+From 2019 to 2023, our internal microservices have seen substantial growth. During this period, the Kitex framework has undergone numerous cycles of optimization and testing to enhance its performance and efficiency. In this article, we share performance optimization techniques that we've systematically implemented over the past few years.
```
Comment (GuangmingLuo) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:18]: adding some external link for Kitex will be better ?

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md (hunk 1 above)
Comment (yy2so) [reply content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md]: My bad, forgot to update the header section, thanks for the review

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md (hunk 2 above)
Comment (yy2so) [reply content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md]: Sure, done

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+2. There's a constraint around developmental efficiency: the growth of the code repository, team size, and code complexity do not have a linear relationship. This makes maintenance incrementally more challenging as the business grows, resulting in a more complicated online implementations.
+
+#### The Shift: Microservice Architecture Era
+
+To overcome the issues inherent in the monolithic architecture, the IT community embarked on a journey into the era of microservice architecture. Here's an example of typical code used in a microservice architecture:
+
+```Go
+func BuySomething(userId int, itemId int) {
+user := client.GetUser(userId) // RPC call
+sth := client.GetItem(itemId) // RPC call
```
Comment (felix021) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:67]: wrong indent

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 4]
```diff
@@ -0,0 +1,538 @@
...
+
+3. When a struct undergoes NoCopy serialization, the output typically takes the form of a two-dimensional byte array. However, Go's `Write([]byte)` interface falls short as it does not support handling non-continuous memory data.
+
+4. Despite being highly compatible, it's provided by Go Runtime and not conducive or suitable for adding new features. 
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/14.jpeg)
+
+#### Go Coding Example
+
+```Go
```
Comment (felix021) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:259]: Not placed correctly according to the original doc.

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 5]
```diff
@@ -0,0 +1,538 @@
...
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/19.jpeg)
+
+In pursuit of enhancing the efficiency of inter-process communication, we developed a communication mode based on shared memory. Shared memory communication throws in the complexity of managing the synchronization of various communication states across different processes. 
+
+To tackle this, we utilized our communication protocol and retained **UDS** as the event notification channel (IO Queue) and shared memory as the data transmission channel (Buffer).
+
+![Image](/img/blog/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/20.jpeg)
+
+For a detailed technical understanding of shmipc, you can refer to our previously published article: [_ByteDance Open Source Shmipc: High Performance IPC Based on Shared Memory_](https://www.cloudwego.io/blog/2023/04/04/introducing-shmipc-a-high-performance-inter-process-communication-library/).
```
Comment (felix021) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:343]: The title is not consistent with that of the link

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+
+* **Alter the Concurrency of Requests in the Code**: It's essential for businesses to iteratively test and evaluate the trade-off between the latency gains from increasing concurrency and the stability loss at peak levels to determine an optimal concurrency value.
+
+* **Use Batch Interfaces**: If the business scenario allows, replacing the current interface with a batch interface could be a more effective strategy.
+
+# Looking into the Future of Optimization
+
+## The Final Frontier: Kernel
+
+The only area currently open for further optimization is the Kernel.
```
Comment (felix021) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:496]: The translations is not exactly what it means in the original better.

The following is closer:

"Currently the only area we haven't explored for optimization is the Kernel."

The areas we discuss before may have room for further optimization and we're still working on some of them.

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md [hunk 7]
```diff
@@ -0,0 +1,538 @@
...
+
+**Netpoll Network Library:**
+
+* Refactoring interfaces to ensure seamless integration with existing libraries in the Go ecosystem.
+
+* Implementing support for Shared Memory Communications over RDMA (SMC-R).
+
+**Pod Affinity:**
+
+* Expanding from same-machine to same-rack granularity, effectively reducing network latency and improving performance.
```
Comment (felix021) [inline content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:538]: The ending is a bit abrupt.

Might be better with a little more like:

> Above are some of our thoughts which we hope may be constructive to you. Should you have some good ideas, you're welcome to discuss with us.

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md (hunk 6 above)
Comment (yy2so) [reply content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md]: Thanks!

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md (hunk 7 above)
Comment (yy2so) [reply content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:538]: I added a conclusion

File: content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md (hunk 4 above)
Comment (yy2so) [reply content/en/blog/news/Enhancing_Performance_in_Microservice_Architecture_with_Kitex/Enhancing_Performance_in_Microservice_Architecture_with_Kitex.md:259]: I moved and changed a few things. If the code and logic are correct, it's better to leave it like this. It's not a 1 to 1 translation

**Review by felix021 — APPROVED** [review]

**Pull request merged**